### PR TITLE
Use flywheel-io/healthcare-api client

### PR DESCRIPTION
**Pull request by davidfarkas**

### CODE REVIEW

File: .coverage [hunk 1]
```diff
@@ -0,0 +1 @@
+!coverage.py: This is a private format, don't read it directly!{"lines":{"/src/run.py":[3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,27,31,32,33,34,38,39,40,44,92,136,187,242,251,264,279,290,315,325,338,351,352,356,361,362,387,395,396,461,471,481,493,93,94,96,97,98,99,100,101,102,105,106,108,109,111,112,114,115,116,118,119,120,121,123,124,125,127,129,130,131,132,133,363,364,365,366,368,388,389,390,370,371,372,373,374,375,376,377,379,380,381,382,384,385,137,138,139,140,142,143,146,147,148,149,150,153,154,156,265,266,267,269,275,268,270,271,272,276,273,157,159,316,317,318,319,320,321,322,160,161,162,163,166,167,168,172,173,174,176,177,179,180,181,182,183,184,188,189,190,191,193,194,397,398,399,401,402,405,406,409,410,412,414,417,418,419,420,421,403,424,462,463,464,466,469,425,426,472,473,475,476,477,479,427,430,431,433,434,435,436,439,440,441,442,443,444,445,446,448,482,483,484,485,486,487,488,489,450,451,452,453,454,455,456,457,458,197,198,199,200,201,204,206,207,209,210,211,212,213,214,215,218,219,220,224,227,228,229,231,232,234,339,340,342,235,236,237,238,239,45,46,47,48,49,50,52,54,56,58,61,62,63,64,65,66,67,68,71,72,73,75,76,77,78,80,81,83,84,86,87,89,90]}}
```
Comment (davidfarkas) [inline .coverage]: please "git"ignore `.coverage` file

File: Dockerfile [hunk 2]
```diff
@@ -1,21 +1,17 @@
...
-COPY manifest.json "${GEAR_MANIFEST_FILE}"
-COPY run.sh "${GEAR_ENTRYPOINT}"
-COPY script.py "${GEAR_BASE_DIR}/script.py"
+COPY . .
 
-WORKDIR ${GEAR_BASE_DIR}
+ENTRYPOINT ["/flywheel/v0/run.py"]
+
+FROM base as testing
+RUN pip install -r tests/requirements.txt
```
Comment (davidfarkas) [inline Dockerfile]: missing new line at the end of file

File: tests/bin/docker-test.sh [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+unset CDPATH
+cd "$( dirname "$0" )/../.."
+
+USAGE="
+Usage:
+    $0 [OPTION...] [[--] TEST_ARGS...]
+Run tests in a docker container.
+Options:
+    -h, --help      Print this help and exit
+    -B, --no-build      Skip building reaper image
```
Comment (davidfarkas) [inline tests/bin/docker-test.sh]: change `reaper`

File: tests/bin/tests.sh [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+set -eu
+unset CDPATH
+cd "$( dirname $0 )/../.."
+
+
+USAGE="
+Usage:
+    $0 [OPTION...] [[--] PYTEST_ARGS...]
+Runs all tests and linting.
+Assumes running in test container or that flywheel_migration and all of its
```
Comment (davidfarkas) [inline tests/bin/tests.sh]: flywheel_migration?

File: manifest.json [hunk 5]
```diff
@@ -1,58 +1,69 @@
 {
 	"name": "ghc-import",
 	"label": "GHC import",
-	"description": "Import dicom files from Google Healthcare API",
-	"version": "0.2.9",
+	"description": "Import dicoms from Google Healthcare API",
```
Comment (davidfarkas) [inline manifest.json]: Use uppercase `DICOM`'s here too as you did elsewhere

**Review by davidfarkas — COMMENTED** [review]
Some minor comments, otherwise LGTM. (tried to only check @denesd's changes )

File: Dockerfile [hunk 6]
```diff
@@ -1,21 +1,17 @@
-FROM python:2.7-alpine3.7
+FROM python:3.7.4-alpine3.10 as base
 
-ENV GEAR_BASE_DIR=/flywheel/v0
-ENV GEAR_INPUT_DIR="${GEAR_BASE_DIR}/input" \
-    GEAR_OUTPUT_DIR="${GEAR_BASE_DIR}/output" \
-    GEAR_MANIFEST_FILE="${GEAR_BASE_DIR}/manifest.json" \
-    GEAR_ENTRYPOINT="${GEAR_BASE_DIR}/run"
+RUN apk add --no-cache bash git \
+    && rm -rf /var/cache/apk/*
```
Comment (ambrussimon) [inline Dockerfile]: Doesn't `--no-cache` already take care of this?

File: .dockerignore [hunk 7]
```diff
@@ -0,0 +1,8 @@
+**
+
+!script.py
```
Comment (ambrussimon) [inline .dockerignore]: This seems outdated.

File: Dockerfile [hunk 8]
```diff
@@ -1,21 +1,17 @@
...
-RUN apk add --no-cache git bash build-base python-dev py-pip jpeg-dev zlib-dev \
-    && mkdir -p "${GEAR_INPUT_DIR}" \
-    && mkdir -p "${GEAR_OUTPUT_DIR}"
+WORKDIR /flywheel/v0
+COPY requirements.txt requirements.txt
 
-# install requirements
-COPY requirements.txt "${GEAR_BASE_DIR}/requirements.txt"
-RUN pip install -r "${GEAR_BASE_DIR}/requirements.txt"
+RUN pip3 install -r requirements.txt \
```
Comment (ambrussimon) [inline Dockerfile:7]: I would suggest keeping the copying of reqs and their installation close together (ie. no empty line).

File: manifest.json [hunk 9]
```diff
@@ -1,57 +1,67 @@
...
 	"label": "GHC import",
-	"description": "Import dicom files from Google Healthcare API",
-	"version": "0.2.9",
+	"description": "Import DICOMs from Google Healthcare API",
+	"version": "0.5.1",
 	"flywheel": "0",
 	"inputs": {
+		"import_ids": {
+			"base": "file",
+			"description": "JSON file with DICOM Series Instance or Study Instance UIDs, HL7 Message IDs , FHIR Resource References. Schema: {'dicoms': ['1.2.840.113619.2.243.4231785106118302.10626.7104.1396227'], 'fhirs': ['Patient/d879d684-892a-4ba7-c4c0-9b68c481635f', 'Encounter/43dc1f3a-049b-49d0-b67e-21cd64acb1d9'], 'hl7s': ['sXiWf0k3rtURTkhi7144lsgfWgbP41OG-3fv5zvjLtM=']}",
```
Comment (ambrussimon) [inline manifest.json]: Study/SeriesInstanceUIDs. Could you check on other gears, maybe there's an already established (nicer) way of handling longer descriptions. Also, `import_ids` is not very descriptive as it is (not that I have an alternative...)

File: run.py [hunk 10]
```diff
@@ -0,0 +1,497 @@
...
+import datetime
+import json
+import logging
+import os
+import pprint
+import shutil
+import sys
+import tempfile
+import zipfile
+import re
```
Comment (ambrussimon) [inline run.py]: Imports are not in order.

File: run.py [hunk 11]
```diff
@@ -0,0 +1,497 @@
...
+from healthcare_api.client import Client as HealthcareAPIClient
+import pytz
+import requests
+from urllib.parse import urljoin
+from dicomweb_client.api import load_json_dataset
+from flywheel_migration.dcm import DicomFile
+from flywheel_migration.util import DEFAULT_TZ
+from requests_toolbelt.multipart.encoder import MultipartEncoder
+
+log = logging.getLogger('ghc_import')
```
Comment (ambrussimon) [inline run.py:27]: Gear contexts have helpers for setting up logging, too.

File: run.py [hunk 12]
```diff
@@ -0,0 +1,497 @@
...
+        return obj.isoformat()
+    elif isinstance(obj, datetime.tzinfo):
+        return obj.zone
+    elif hasattr(obj, 'encode'):
+        return str(obj.encode())
+
+    raise TypeError(repr(obj) + ' is not JSON serializable')
+
+
+class FwApi(requests.Session):
```
Comment (ambrussimon) [inline run.py:343]: You may as well pass the key and adjust the headers in the constructor IMO.

File: tests/requirements.txt [hunk 13]
```diff
@@ -0,0 +1,3 @@
+pytest
```
Comment (ambrussimon) [inline tests/requirements.txt]: Requirements are not in order.

File: tests/bin/docker-test.sh [hunk 14]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env sh
```
Comment (ambrussimon) [inline tests/bin/docker-test.sh]: The `docker-tests` and `tests.sh` pattern was ported rather hastily and lazily.
There's a bunch of lines that were just copied without any thought to it (`see tests.sh` L# 28-34).
Test requirements aren't heavy at all - I recommend just
- removing these wrappers
- requiring users to install the requirements and run pytest locally
- automating the same for CI

**Review by ambrussimon — CHANGES_REQUESTED** [review]

File: run.py [hunk 15]
```diff
@@ -0,0 +1,492 @@
...
+    'U': 'Unknown or Not Reported',
+}
+
+
+def main(context):
+    config = context.config
+
+    log.setLevel(getattr(logging, config['log_level']))
+
+    log.debug('config.json\n%s', pprint.pformat(config))
```
Comment (ambrussimon) [inline run.py]: Is this logging more than what's available via `context.log_config()`?

File: run.py [hunk 16]
```diff
@@ -0,0 +1,492 @@
...
+import shutil
+import sys
+import tempfile
+import zipfile
+from dicomweb_client.api import load_json_dataset
+from flywheel_migration.dcm import DicomFile
+from flywheel_migration.util import DEFAULT_TZ
+from healthcare_api.client import Client as HealthcareAPIClient
+from requests_toolbelt.multipart.encoder import MultipartEncoder
+from urllib.parse import urljoin
```
Comment (ambrussimon) [inline run.py]: Minor complaint here: `urllib` is in core, `pytz` and `requests` isn't, and there's no empty line dividing core/3rd-party imports.

**Review by ambrussimon — APPROVED** [review]
Well done, this looks good to me!

File: run.py [hunk 17]
```diff
@@ -0,0 +1,492 @@
...
+        host_ip = '192.168.50.162'
+        with open('/etc/hosts', 'a') as f:
+            f.write(host_ip + '\tdocker.local.flywheel.io\n')
+    api_uri = api_key.rsplit(':', 1)[0]
+    if not api_uri.startswith('http'):
+        api_uri = 'https://' + api_uri + '/api'
+    fw_api = FwApi(api_uri, api_key)
+
+    # validate destination project exists
+    resp = fw_api.get('projects/' + config['project_id'])
```
Comment (davidfarkas) [inline run.py:63]: I think we should minimize the usage of custom api clients. I suppose that we can use the SDK to validate that the project exists or not.
@ambrussimon any thoughts?

Comment (davidfarkas) [reply run.py:63]: Never mind, I will open a PR soon which includes this change.

**Pull request merged**